### PR TITLE
New rule: tabpanel should be related to a tab

### PR DESCRIPTION
**Pull request by ctaymor**
per the recomendations from the WAI-ARIA guidelines on the role tabpanel
found here: http://www.w3.org/TR/wai-aria/roles#tabpanel

Addresses #228

### CODE REVIEW

**Comment by ctaymor** [comment]
And we can add this to the Wiki:

## AX_ARIA_13
##

**Good:**
```html
<!-- Good: each tabpanel is referenced by a aria-controls on a different tab -->
<div role="tablist">
  <div role="tab" aria-controls="tabpanel1">Tab 1</div>
  <div role="tab" aria-controls="tabpanel2">Tab 2</div>
</div>

<div role="tabpanel" id="tabpanel1"> Some content in the first tab panel</div>
<div role="tabpanel" id="tabpanel2"> Other stuff in the second tab panel</div>

<!-- Good: each tabpanel has an aria-labelledby for a different tab -->
<div role="tablist">
  <div role="tab" id="tab1">Tab 1</div>
  <div role="tab" id="tab2">Tab 2</div>
</div>

<div role="tabpanel" aria-labeledby="tab1"> Some content in the first tab panel</div>
<div role="tabpanel" aria-labeledby="tab2"> Other stuff in the second tab panel</div>
```

**Bad:**
```html
<!-- Bad: tabpanel is not labeled by and there is no tab which controls it -->
<div role="tablist">
  <div role="tab">Tab 1</div>
  <div role="tab">Tab 2</div>
</div>

<div role="tabpanel"> Some content in the first tab panel</div>
<div role="tabpanel"> Other stuff in the second tab panel</div>

<!-- Bad: tabpanel is labeled by something other than a tab and not controlled by a tab -->
<div role="tablist">
  <div role="tab">Tab 1</div>
</div>

<div role="tabpanel" aria-labeledby="notATab"> Some content in the first tab panel</div>

<h5 id="notATab">This describes the content of the tabpanel</h5>

```

An element with the `tabpanel` ARIA roles should be related to an element with the role `tab`. Either the tabpanel can have an `aria-labelledby` referencing the tab's id, or the tab can have an `aria-controls` referencing the tabpanel's id.

You can find more information on the tabpanel role on the [WAI-ARIA site](http://www.w3.org/TR/wai-aria/roles#tabpanel).

**Comment by dylanb** [comment]
This will generate mostly false positives for two reasons

1) it is only a SHOULD requirement, and
2) the tab panels can be added dynamically by JavaScript and only have to be there when the tab itself is selected

**Comment by ctaymor** [comment]
The second concern shouldn't be a problem with this audit. It only checks that the tab panels have a tab which controls them, not whether the tab has a tab panel. So if the tab panel hasn't been added yet, it won't fail the test.

In regards to the first concern, I gave it a warning severity instead of severe for that reason. However, my understanding is that although this is only a SHOULD requirement, it is of great importance for the usability of tabs, and thus worth having an audit for.

File: test/audits/uncontrolled-tabpanel-test.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  var tab = document.createElement('div');
+  tab.setAttribute('role', 'tab');
+  tab.setAttribute('aria-controls', 'tabpanelId');
+
+  tabList.appendChild(tab);
+  var tabPanel = document.createElement('div');
+  tabPanel.setAttribute('role', 'tabpanel');
+  tabPanel.setAttribute('id', 'tabpanelId');
+  fixture.appendChild(tabPanel);
+
```
Comment (dylanb) [inline test/audits/uncontrolled-tabpanel-test.js]: Should add a passing test where there is more than one tab, but only one tabpanel to ensure the false positives do not occur

**Comment by ctaymor** [comment]
Good idea @dylanb. I added the test for that case (and caught a bug in the process).

File: src/audits/UncontrolledTabpanel.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+goog.require('axs.AuditRules');
+goog.require('axs.browserUtils');
+goog.require('axs.constants.Severity');
+
+function labeledByATab(element) {
```
Comment (alice) [inline src/audits/UncontrolledTabpanel.js]: Similarly to #218, could you scope these methods to this file and JSDoc them?

File: test/audits/uncontrolled-tabpanel-test.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+module('UncontrolledTabpanel');
+
+test('No elements === no problems.', function() {
```
Comment (alice) [inline test/audits/uncontrolled-tabpanel-test.js]: I think this tab and the one below are redundant with the one on line 38.

Comment (alice) [inline test/audits/uncontrolled-tabpanel-test.js]: Hm, actually, I see the value in the one below, but this one is certainly redundant.

File: Changelog.md [hunk 4]
```diff
@@ -59,6 +59,7 @@
 * Elements with ARIA roles must ensure required owned elements are present (`src/audits/RequiredOwnedAriaRoleMissing.js`)
 * Avoid positive integer values for tabIndex (`src/audits/TabIndexGreaterThanZero.js`)
 * This element has an unsupported ARIA attribute (`src/audits/UnsupportedAriaAttribute.js`)
+* A tabpanel should be related to a tab via aria-controls or aria-labelledby (`src/audits/UncontrolledTabpanel.js)
```
Comment (alice) [inline Changelog.md]: As with #218, we'll need to move this Changelog entry up above the latest release line (sorry!)

**Comment by alice** [comment]
:+1: once comments addressed and Changelog updated.

**Comment by ctaymor** [comment]
Thanks Alice. I made those changes.

**Comment by alice** [comment]
@ctaymor Did you update the Changelog too? You may need to pull the latest version.

**Comment by ctaymor** [comment]
@alice Sorry, I did update the changelog, but not correctly because I hadn't pulled in the latest version. This should fix it with the latest changelog.

**Comment by alice** [comment]
Thanks!

**Pull request merged**